pkg/kubelet/cm/topologymanager/policy_single_numa_node.go

```diff
@@ -52,8 +54,122 @@ func (p *singleNumaNodePolicy) canAdmitPodResult(hint *TopologyHint) lifecycle.P
+	// Loop through all provider hints and save an accumulated list of the
+	// hints returned by each hint provider. If no hints are provided, assume
+	// that provider has no preference for topology-aware allocation.
+	var allResourcesHints [][]TopologyHint
+	for _, hints := range providersHints {
+		if len(hints) == 0 {
+			klog.Infof("[topologymanager] Hint Provider has no preference for NUMA affinity with any resource, " +
+				"skipping.")
+			continue
+		}
+
+		// Otherwise, accumulate the hints for each resource type into allProviderHints.
+		for resource := range hints {
+			if hints[resource] == nil {
+				klog.Infof("[topologymanager] Hint Provider has no preference for NUMA affinity with resource "+
+					"'%s', skipping.", resource)
+				continue
+			}
+
+			if len(hints[resource]) == 0 {
+				klog.Infof("[topologymanager] Hint Provider has no possible NUMA affinities for resource '%s'",
+					resource)
+				// return defaultHint which will fail pod admission
+				return TopologyHint{}
+			}
+			klog.Infof("[topologymanager] TopologyHints for resource '%v': %v", resource, hints[resource])
+			allResourcesHints = append(allResourcesHints, hints[resource])
+		}
+	}
+	// In case allProviderHints length is zero it means that we have no
+	// preference for NUMA affinity. In that case update default hint preferred
+	// to true to allow scheduling.
+	if len(allResourcesHints) == 0 {
+		klog.Infof("[topologymanager] No preference for NUMA affinity from all providers")
+		return TopologyHint{nil, true}
+	}
+
+	allResourcesHints = p.filterHints(allResourcesHints)
+	// If no hints, or there is a resource with empty hints after filtering, then policy
+	// cannot be satisfied
+	if len(allResourcesHints) == 0 {
+		klog.Infof("[topologymanager] No hints that align to a single NUMA node.")
+		return TopologyHint{}
+	}
+	for _, hints := range allResourcesHints {
+		if len(hints) == 0 {
+			klog.Infof("[topologymanager] No hints that align to a single NUMA node for resource.")
+			return TopologyHint{}
+		}
+	}
+
+	// Set the bestHint to return from this function as an any-NUMANode
+	// affinity with an unpreferred allocation. This will only be returned if
+	// no better hint can be found when merging hints from each hint provider.
+	defaultAffinity, _ := bitmask.NewBitMask(p.numaNodes...)
+	bestHint := TopologyHint{defaultAffinity, false}
+	iterateAllProviderTopologyHints(allResourcesHints, func(permutation []TopologyHint) {
+		mergedHint := p.mergePermutation(permutation)
+		// Only consider mergedHints that result in a NUMANodeAffinity == 1 to
+		// replace the current defaultHint.
+		if mergedHint.NUMANodeAffinity.Count() != 1 {
+			return
+		}
+
+		// If the current defaultHint is the same size as the new mergedHint,
+		// do not update defaultHint
+		if mergedHint.NUMANodeAffinity.Count() == bestHint.NUMANodeAffinity.Count() {
+			return
+		}
```

Comment:
This check seems unnecessary for this policy (given the check right above this). However, I think we need to add a chack to make sure that `bestHint` is always updated to the narrower of the two hints.

Reply:
Without this check, a `bestHint.NUMANodeAffinity` of `{...001}` will be replaced by a `mergedHint.NUMANodeAffinity` of `{...0010}` or `{...0100}` etc as they occur.

As for a check to make sure that `bestHint` is always updated to the narrower of the two hints... `filterHints` ensures that only hints with `NUMANodeAffinity.Count() == 1` are kept... so we should only ever end up with `bestHint` that has `NUMANodeAffinity` of width 0 or 1 - which is handled in the first check.

Reply:
`isNarrowerThan()` handles this case. It chooses the one with the lowest order bits to be considered narrower.
 https://github.com/kubernetes/kubernetes/blob/master/pkg/kubelet/cm/topologymanager/bitmask/bitmask.go#L129

Reply:
Sure, so our three checks could be as follows: (including check for nil `bestHint`)
````
// Only consider mergedHints that result in a NUMANodeAffinity == 1 to
// replace the current bestHint.
if mergedHint.NUMANodeAffinity.Count() != 1 {
       return
}

// If the current bestHint is nil, replace with mergedHint.
if bestHint == (TopologyHint{}) {
       bestHint = mergedHint
       return
}

// Only consider mergedHints that have a narrower NUMANodeAffinity
// than the NUMANodeAffinity in the current bestHint.
if !mergedHint.NUMANodeAffinity.IsNarrowerThan(bestHint.NUMANodeAffinity) {
     return
}

Reply:
Looks good. Only slight modification I'd make is to check `if bestHint.NUMANodeAffinity == nil` instead of `bestHint == (TopologyHint{})` since it's the fact that that field is `nil` that we have to protect against.